Emit schema with all model properties optional + allow partial updates via PATCH

bajtos opened the pull request:
- Allow Controller methods implementing PATCH operation to describe their input data as "a model with all properties optional". To emit schema with optional model properties, we are adding a new schema-generation option called `partial`.
- Modify CLI templates to use the new `partial` option to describe request body parameter.
- Modify example applications to leverage this new option too.
- Update code snippets in our docs for consistency.

See https://github.com/strongloop/loopback-next/issues/2652, https://github.com/strongloop/loopback-next/issues/1722 and https://github.com/strongloop/loopback-next/issues/1179

<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

b-admike commented on packages/repository-json-schema/src/build-schema.ts:
```diff
@@ -16,10 +16,44 @@ import {JSONSchema6 as JSONSchema} from 'json-schema';
+      if (options.includeRelations) return MODEL_TYPE_KEYS.ModelWithRelations;
+  }
+
+  const flags = Object.keys(options);
```
nitpick: can we define flags before the switch block and use it throughout?

bajtos replied:
Do you meant to use `flags.length` instead of `Object.keys(options).length`? Is there any other place where to use `flags`?

b-admike replied:
Yea that's right 👍

b-admike commented on packages/repository-json-schema/src/build-schema.ts:
```diff
@@ -16,10 +16,44 @@ import {JSONSchema6 as JSONSchema} from 'json-schema';
+ * @private
+ */
+export function buildModelCacheKey(
+  ctor: Function,
```
Is there a reason for the `ctor` parameter in this function? I don't see it being used.

bajtos replied:
Good catch, I'll remove it.

samarpanB commented on packages/repository-json-schema/src/keys.ts:
```diff
@@ -15,6 +19,6 @@ export const enum MODEL_TYPE_KEYS {
  * Metadata key used to set or retrieve repository JSON Schema
  */
 export const JSON_SCHEMA_KEY = MetadataAccessor.create<
-  {[options in MODEL_TYPE_KEYS]: JSONSchema},
```
Any specific reason for removing this type check ?

bajtos replied:
In the future, we will be adding more schema options like `exclude` (see https://github.com/strongloop/loopback-next/issues/2653). Because options can be combined, e.g. `partial` with `exclude`, the number of possible combinations (and thus cache) keys is growing fast. I find it impractical to enumerate all valid keys in a TypeScript enum type. We will need 4 entries now, 8 entries when we implement `exclude`, 16 entries when we introduce another option, etc.

That's why I switched the indexer type from `options in MODLE_TYPE_KEYS` to `key: string`.

samarpanB replied:
Makes sense.

samarpanB commented on packages/repository-json-schema/src/build-schema.ts:
```diff
@@ -16,10 +16,44 @@ import {JSONSchema6 as JSONSchema} from 'json-schema';
+    case 0:
+      return MODEL_TYPE_KEYS.ModelOnly;
+    case 1:
+      if (options.includeRelations) return MODEL_TYPE_KEYS.ModelWithRelations;
```
Don’t you think cyclomatic complexity increases with if inside switch? Just a thought.

bajtos replied:
Good point, I am going to simplify this function to make it easier to understand and reason about. Thank you for pointing it out! ❤️

samarpanB commented on examples/todo-list/src/controllers/todo.controller.ts:
```diff
@@ -89,7 +90,14 @@ export class TodoController {
+        },
+      },
+    })
+    todo: Partial<TodoList>,
```
This should be ToDO

nabdelgadir commented on examples/todo-list/src/controllers/todo-list.controller.ts:
```diff
@@ -80,7 +81,14 @@ export class TodoListController {
+        },
+      },
+    })
+    obj: Partial<TodoList>,
```
nitpick: `obj` -> `todoList`

nabdelgadir commented on examples/todo-list/src/controllers/todo-list.controller.ts:
```diff
@@ -108,7 +116,14 @@ export class TodoListController {
+        },
+      },
+    })
+    obj: Partial<TodoList>,
```
nitpick: `obj` -> `todoList`

nabdelgadir commented on examples/todo-list/src/controllers/todo.controller.ts:
```diff
@@ -89,7 +90,14 @@ export class TodoController {
+    @requestBody({
+      content: {
+        'application/json': {
+          schema: getModelSchemaRef(TodoList, {partial: true}),
```
`getModelSchemaRef(TodoList, {partial: true}),` -> `getModelSchemaRef(Todo, {partial: true}),`

nabdelgadir commented on packages/repository-json-schema/src/keys.ts:
```diff
@@ -6,6 +6,10 @@
+/**
+ * TODO(semver-major) remove these constants in the next major version
+ * @deprecated
+ */
```
Can I ask why they're being removed?

bajtos replied:
Of course you can! 😄 

Cross-posting my comment from the discussion thread https://github.com/strongloop/loopback-next/pull/3199#discussion_r296590295:

> In the future, we will be adding more schema options like `exclude` (see https://github.com/strongloop/loopback-next/issues/2653). Because options can be combined, e.g. `partial` with `exclude`, the number of possible combinations (and thus cache) keys is growing fast. I find it impractical to enumerate all valid keys in a TypeScript enum type. We will need 4 entries now, 8 entries when we implement `exclude`, 16 entries when we introduce another option, etc.
>
> That's why I switched the indexer type from `options in MODLE_TYPE_KEYS` to `key: string`.

I am going to improve this comment to tell developers to use `buildModelCacheKey` instead.

raymondfeng commented on packages/repository-json-schema/src/build-schema.ts:
```diff
@@ -16,10 +16,44 @@ import {JSONSchema6 as JSONSchema} from 'json-schema';
+   * Set this flag to mark all model properties as optional. This is typically
+   * used to describe request body of PATCH endpoints.
+   */
+  partial?: boolean;
```
In the future, we can support more options such as `exclude`.

bajtos replied:
Definitely, see e.g. https://github.com/strongloop/loopback-next/issues/2653:

> Allow Controller methods implementing CREATE operation to describe their input data as "a model without `id` and `_rev` properties ". This story requires #2629 and #2631 to be implemented first.
> 
> To emit schema with certain properties excluded, we should add a new schema-generation option called `exclude` & accepting a list of property names to remove from the schema. To allow TypeScript compiler to verify that `exclude` items are valid property names, we should leverage TypeScript generics and `keyof` keyword.

bajtos commented:
@b-admike @nabdelgadir @raymondfeng @samarpanB Thank you for the review. I believe I have addressed your comments in 6e167805a and 59f0dbc67. LGTY now?

nabdelgadir commented:
> Should we perhaps use `+` instead of `;`?
> 
> ```ts
> expect(key).to.equal('includeRelations+partial');
> ```

I like `+` more than `;` personally
